README


This archive contains the distribution Parse-Keyword,
version 0.08:

  DEPRECATED: write syntax extensions in perl

This software is Copyright (c) 2013 by Jesse Luehrs.

This is free software, licensed under:

  The MIT (X11) License




Changes
Revision history for Parse-Keyword

0.08  2013-10-10
      - doc tweaks to make the deprecation notice more visible

0.07  2013-10-10
      - THIS MODULE IS NO LONGER SUPPORTED DUE TO UNFIXABLE BROKENNESS
        see the warning in the documentation for more details
      - fix tests for newer Carp (abraxxa)

0.06  2013-08-01
      - support building on windows (forwardever, #3)

0.05  2013-07-24
      - fix skipping tests

0.04  2013-07-23
      - pass the keyword name into the parsing function

0.03  2013-07-23
      - the parse_* functions now return undef when there was a parse error
      - various bugs related to error handling should be fixed
      - remove unnecessary test dep on B::Hooks::EndOfScope

0.02  2013-07-22
      - work around some issues that would happen when you peeked past the end
        of the line
      - allow parsing into non-anonymous subs (this makes a difference in some
        obscure situations)
      - fix compilation on threaded perls

0.01  2013-07-21
      - Initial release


callparser1.h
/* DO NOT EDIT -- generated by Devel::CallParser version 0.001 */
#ifndef C8K61oRQKxigiqmUlVdk_INCLUDED_callparser1
#define C8K61oRQKxigiqmUlVdk_INCLUDED_callparser1 1
#ifndef PERL_VERSION
 #error you must include perl.h before callparser1.h
#elif !(PERL_REVISION == 5 && PERL_VERSION == 18 && PERL_SUBVERSION == 0)
 #error this callparser1.h is for Perl 5.18.0 only
#endif /* Perl version mismatch */
PERL_CALLCONV OP * C8K61oRQKxigiqmUlVdk_pac0(pTHX_ U32 *);
#define Perl_parse_args_parenthesised C8K61oRQKxigiqmUlVdk_pac0
#define parse_args_parenthesised(fp) Perl_parse_args_parenthesised(aTHX_ fp)
PERL_CALLCONV OP * C8K61oRQKxigiqmUlVdk_paz0(pTHX_ U32 *);
#define Perl_parse_args_nullary C8K61oRQKxigiqmUlVdk_paz0
#define parse_args_nullary(fp) Perl_parse_args_nullary(aTHX_ fp)
PERL_CALLCONV OP * C8K61oRQKxigiqmUlVdk_pau0(pTHX_ U32 *);
#define Perl_parse_args_unary C8K61oRQKxigiqmUlVdk_pau0
#define parse_args_unary(fp) Perl_parse_args_unary(aTHX_ fp)
PERL_CALLCONV OP * C8K61oRQKxigiqmUlVdk_pal0(pTHX_ U32 *);
#define Perl_parse_args_list C8K61oRQKxigiqmUlVdk_pal0
#define parse_args_list(fp) Perl_parse_args_list(aTHX_ fp)
PERL_CALLCONV OP * C8K61oRQKxigiqmUlVdk_pab0(pTHX_ U32 *);
#define Perl_parse_args_block_list C8K61oRQKxigiqmUlVdk_pab0
#define parse_args_block_list(fp) Perl_parse_args_block_list(aTHX_ fp)
PERL_CALLCONV OP * C8K61oRQKxigiqmUlVdk_pap0(pTHX_ GV *, SV *, U32 *);
#define Perl_parse_args_proto C8K61oRQKxigiqmUlVdk_pap0
#define parse_args_proto(gv, sv, fp) Perl_parse_args_proto(aTHX_ gv, sv, fp)
PERL_CALLCONV OP * C8K61oRQKxigiqmUlVdk_pan0(pTHX_ GV *, SV *, U32 *);
#define Perl_parse_args_proto_or_list C8K61oRQKxigiqmUlVdk_pan0
#define parse_args_proto_or_list(gv, sv, fp) Perl_parse_args_proto_or_list(aTHX_ gv, sv, fp)
typedef OP *(*Perl_call_parser)(pTHX_ GV *, SV *, U32 *);
#define CALLPARSER_PARENS    0x00000001
#define CALLPARSER_STATEMENT 0x00000002
PERL_CALLCONV void C8K61oRQKxigiqmUlVdk_gcp1(pTHX_ CV *, Perl_call_parser *, SV **);
#define Perl_cv_get_call_parser C8K61oRQKxigiqmUlVdk_gcp1
#define cv_get_call_parser(cv, fp, op) Perl_cv_get_call_parser(aTHX_ cv, fp, op)
PERL_CALLCONV void C8K61oRQKxigiqmUlVdk_scp1(pTHX_ CV *, Perl_call_parser, SV *);
#define Perl_cv_set_call_parser C8K61oRQKxigiqmUlVdk_scp1
#define cv_set_call_parser(cv, f, o) Perl_cv_set_call_parser(aTHX_ cv, f, o)
#endif /* !C8K61oRQKxigiqmUlVdk_INCLUDED_callparser1 */


lib/Parse/Keyword.pm
package Parse::Keyword;
BEGIN {
  $Parse::Keyword::AUTHORITY = 'cpan:DOY';
}
{
  $Parse::Keyword::VERSION = '0.07';
}
use strict;
use warnings;
use 5.014;
# ABSTRACT: DEPRECATED: write syntax extensions in perl

use Devel::CallParser;
use XSLoader;

XSLoader::load(
    __PACKAGE__,
    exists $Parse::Keyword::{VERSION} ? ${ $Parse::Keyword::{VERSION} } : (),
);



sub import {
    my $package = shift;
    my ($keywords) = @_;

    my $caller = caller;

    for my $keyword (keys %$keywords) {
        my $sub = do {
            no strict 'refs';
            \&{ $caller . '::' . $keyword };
        };
        install_keyword_handler($sub, $keywords->{$keyword});
    }

    my @helpers = qw(
        lex_peek
        lex_read
        lex_read_space
        lex_stuff
        parse_block
        parse_stmtseq
        parse_fullstmt
        parse_barestmt
        parse_fullexpr
        parse_listexpr
        parse_termexpr
        parse_arithexpr
        compiling_package
    );

    for my $helper (@helpers) {
        no strict 'refs';
        *{ $caller . '::' . $helper } = \&{ __PACKAGE__ . '::' . $helper };
    }
}


1;

__END__

=pod

=head1 NAME

Parse::Keyword - DEPRECATED: write syntax extensions in perl

=head1 VERSION

version 0.07

=head1 SYNOPSIS

  use Parse::Keyword { try => \&try_parser };
  use Exporter 'import';
  our @EXPORT = 'try';

  sub try {
      my ($try, $catch) = @_;
      &Try::Tiny::try($try, ($catch ? (&Try::Tiny::catch($catch)) : ()));
  }

  sub try_parser {
      lex_read_space;
      die "syntax error" unless lex_peek eq '{';
      my $try = parse_block;
      lex_read_space;

      my $catch;
      if (lex_peek(6) =~ /^catch\b/) {
          lex_read(5);
          lex_read_space;
          die "syntax error" unless lex_peek eq '{';
          $catch = parse_block;
      }

      return (sub { ($try, $catch) }, 1);
  }

=head1 DESCRIPTION

B<< NOTE: The API of this module is still in flux. I may make
backwards-incompatible changes as I figure out how it should look. >>

This module allows you to write keyword-based syntax extensions without
requiring you to write any C code yourself. It is similar to L<Devel::Declare>,
except that it uses the Perl parser API introduced in Perl 5.14 in order to
allow you to parse parts of things using perl's own parser, rather than having
to fake it with balanced brace matching or other fragile things.

To use this module, you should pass a hashref to the C<use> statement. The keys
of this hashref are subroutines in the current package which should have
special parsing behavior attached to them, and the values are coderefs which
should be used to implement the custom parsing behavior.

The parsing coderefs will be called when perl encounters a call to the keyword
that you attached custom parsing to. The current parser state will be directly
after parsing the keyword. The parser function will receive the name of the
keyword as a parameter, and should return a coderef which, when called at
runtime, will produce the arguments to the function. In addition, if your
keyword should be parsed as a statement (for instance, if you don't want to
require a trailing semicolon), you can return a second, true value.

In order to actually handle the parsing itself, this module also exports
various parsing functions, which you can call. See below for details.

=head1 FUNCTIONS

=head2 lex_peek($n)

Returns a string consisting of the next C<$n> characters in the input (or next
one character, if C<$n> isn't given). This string may be shorter than C<$n>
characters if there are fewer than C<$n> characters remaining to read. The
current position in the buffer to be parsed is not moved. See L<<
perlapi/PL_parser->linestr >> and L<perlapi/lex_next_chunk> for more
information.

NOTE: This function currently only returns text that is on the current line,
unless the current line has been fully read (via C<lex_read>). This is due to a
bug in perl itself, and this restriction will hopefully be lifted in a future
version of this module, so don't depend on it. See the L</BUGS> section for
more information.

=head2 lex_read($n)

Moves the current position in the parsing buffer forward by C<$n> characters
(or one character, if C<$n> isn't given). See L<perlapi/lex_read_to> for more
details.

=head2 lex_read_space

Moves the current position in the parsing buffer forward past any whitespace or
comments. See L<perlapi/lex_read_space> for more details.

=head2 lex_stuff($str)

Inserts C<$str> into the current parsing buffer at the current location, so
that future calls to C<lex_peek> and such will see it. Note that this does not
move the current position in the parsing buffer, so multiple calls to
C<lex_stuff> at the same location will end up inserted into the buffer in
reverse order. See L<perlapi/lex_stuff_sv> for more information.

=head2 parse_block, parse_stmtseq, parse_fullstmt, parse_barestmt,
parse_fullexpr, parse_listexpr, parse_termexpr, parse_arithexpr

These functions parse the specified amount of Perl code, and return a coderef
which will evaluate that code when executed. They each take an optional boolean
parameter that should be true if you are creating a subroutine which will be
going in the symbol table, or in other more obscure situations involving
closures (the CVf_ANON flag will be set on the created coderef if this is not
passed - see C<t/unavailable.t> in this distribution). See
L<perlapi/parse_block>, L<perlapi/parse_stmtseq>, L<perlapi/parse_fullstmt>,
L<perlapi/parse_barestmt>, L<perlapi/parse_fullexpr>, L<parse_listexpr>,
L<parse_termexpr>, and L<perlapi/parse_arithexpr> for more details.

=head2 compiling_package

Returns the name of the package that the keyword which is currently being
parsed was called in. This should be used instead of C<caller> if you want to
do something like install a subroutine in the calling package.

=head1 DO NOT USE!

This module has fundamental errors in the way it handles closures, which are
not fixable. Runtime keywords will never be able to work properly with the
current design of this module. There are certain cases where this module is
still safe to use (keywords that only have effect at compile time, or keywords
that never call any of the C<parse_*> functions), but that is limiting enough
to make this module mostly worthless, and I likely won't be continuing to
maintain it. Be warned!

=head1 BUGS

Peeking into the next line is currently (as of 5.19.2) broken in perl if the
current line hasn't been fully consumed. This module works around this by just
not doing that. This shouldn't be an issue for the most part, since it will
only come up if you need to conditionally parse something based on a token that
can span multiple lines. Just keep in mind that if you're reading in a large
chunk of text, you'll need to alternate between calling C<lex_peek> and
C<lex_read>, or else you'll only be able to see text on the current line.

This module also inherits the limitation from L<Devel::CallParser> that custom
parsing is only triggered if the keyword is called by its unqualified name
(C<try>, not C<Try::try>, for instance).

This module doesn't yet work with lexical subs, such as via
L<Exporter::Lexical>. This will hopefully be fixed in the future, but will
likely require modifications to perl.

Please report any bugs to GitHub Issues at
L<https://github.com/doy/parse-keyword/issues>.

=head1 SEE ALSO

L<Devel::CallParser>

L<Keyword::API>

L<Devel::Declare>

=head1 SUPPORT

You can find this documentation for this module with the perldoc command.

    perldoc Parse::Keyword

You can also look for information at:

=over 4

=item * MetaCPAN

L<https://metacpan.org/release/Parse-Keyword>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Parse-Keyword>

=item * Github

L<https://github.com/doy/parse-keyword>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Parse-Keyword>

=back

=for Pod::Coverage   install_keyword_handler

=head1 AUTHOR

Jesse Luehrs <doy@tozt.net>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by Jesse Luehrs.

This is free software, licensed under:

  The MIT (X11) License

=cut


Makefile.PL
# This Makefile.PL for  was generated by Dist::Zilla.
# Don't edit it but the dist.ini used to construct it.
BEGIN { require 5.014; }
use strict;
use warnings;
use ExtUtils::MakeMaker 6.30;

my %WriteMakefileArgs = (
  'ABSTRACT' => 'DEPRECATED: write syntax extensions in perl',
  'AUTHOR' => 'Jesse Luehrs <doy@tozt.net>',
  'BUILD_REQUIRES' => {
    'Carp' => '0',
    'Exporter' => '0',
    'File::Find' => '0',
    'File::Temp' => '0',
    'Test::More' => '0.88',
    'base' => '0',
    'if' => '0',
    'lib' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'Devel::CallParser' => '0',
    'ExtUtils::MakeMaker' => '6.30'
  },
  'DISTNAME' => 'Parse-Keyword',
  'EXE_FILES' => [],
  'LICENSE' => 'mit',
  'NAME' => 'Parse::Keyword',
  'NORECURS' => 1,
  'PREREQ_PM' => {
    'Devel::CallParser' => '0',
    'XSLoader' => '0',
    'strict' => '0',
    'warnings' => '0'
  },
  'VERSION' => '0.08',
  'clean' => {
    'FILES' => 'callparser1.h'
  },
  'test' => {
    'TESTS' => 't/*.t t/fun/*.t t/try/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

use File::Spec::Functions 'abs2rel';
use Devel::CallParser 'callparser1_h', 'callparser_linkable';
open my $fh, '>', 'callparser1.h' or die "Couldn't write to callparser1.h";
$fh->print(callparser1_h);
my @linkable = map { abs2rel($_) } callparser_linkable;
unshift @linkable, '$(BASEEXT)$(OBJ_EXT)' if @linkable;
$WriteMakefileArgs{OBJECT} = join(' ', @linkable) if @linkable;

WriteMakefile(%WriteMakefileArgs);

